### PR TITLE
feat: add task tracker with detailed logging

### PR DESCRIPTION
**Pull request by spacegospod**
**Summary of Pull Request**

We have many long-running operations and we don't display any progress updates.

Users have to rely on the domain manager UI to track progress.

I'm adding a new task tracker that logs the message for each subtask.

It logs each subtask only once, as long as its status is something other than in progress.

To enable the new logging one must set `TF_LOG_PROVIDER_VCF` to a valid log level.

Example:

```console
2024-08-20T18:04:29.957+0300 [INFO]  vcf: esxi-5.vrack.vsphere.local commissionSpec commission initiated. waiting for task id = 877f7868-0b72-4711-a449-b57cec13a63c: tf_resource_type=vcf_host tf_req_id=2c00dec9-6f95-50ed-6a9f-f501143a2252 tf_provider_addr=registry.terraform.io/hashicorp/vcf tf_rpc=ApplyResourceChange
2024-08-20T18:04:40.236+0300 [INFO]  vcf: [SUCCESSFUL] Acquire SDDC Manager System Limited Lock if Resources Exists: tf_provider_addr=registry.terraform.io/hashicorp/vcf tf_rpc=ApplyResourceChange tf_resource_type=vcf_host tf_req_id=2c00dec9-6f95-50ed-6a9f-f501143a2252
2024-08-20T18:04:40.236+0300 [INFO]  vcf: [SUCCESSFUL] Acquire SDDC Manager Resources Lock if Resources exists: tf_req_id=2c00dec9-6f95-50ed-6a9f-f501143a2252 tf_provider_addr=registry.terraform.io/hashicorp/vcf tf_rpc=ApplyResourceChange tf_resource_type=vcf_host
2024-08-20T18:04:40.236+0300 [INFO]  vcf: [SUCCESSFUL] Removes hosts (missing from DB) from orchestrator metadata before retrying a failed commissioning workflow: tf_rpc=ApplyResourceChange tf_resource_type=vcf_host tf_req_id=2c00dec9-6f95-50ed-6a9f-f501143a2252 tf_provider_addr=registry.terraform.io/hashicorp/vcf
2024-08-20T18:04:40.236+0300 [INFO]  vcf: [SUCCESSFUL] Performing Cleanup of SDDC Inventory Before Retrying the Failed ESXi Host Commissioning Workflow: tf_provider_addr=registry.terraform.io/hashicorp/vcf tf_rpc=ApplyResourceChange tf_resource_type=vcf_host tf_req_id=2c00dec9-6f95-50ed-6a9f-f501143a2252
2024-08-20T18:04:40.236+0300 [INFO]  vcf: [SUCCESSFUL] Automation Helper Action: tf_req_id=2c00dec9-6f95-50ed-6a9f-f501143a2252 tf_provider_addr=registry.terraform.io/hashicorp/vcf tf_rpc=ApplyResourceChange tf_resource_type=vcf_host
2024-08-20T18:04:40.236+0300 [INFO]  vcf: [SUCCESSFUL] [Applicable to skipped ESXi Hosts]Perform set of validations for current workflow: tf_rpc=ApplyResourceChange tf_resource_type=vcf_host tf_req_id=2c00dec9-6f95-50ed-6a9f-f501143a2252 tf_provider_addr=registry.terraform.io/hashicorp/vcf
2024-08-20T18:04:40.236+0300 [INFO]  vcf: [SUCCESSFUL] Validate ESXi Hosts Before Starting the Commissioning Process: tf_resource_type=vcf_host tf_req_id=2c00dec9-6f95-50ed-6a9f-f501143a2252 tf_provider_addr=registry.terraform.io/hashicorp/vcf tf_rpc=ApplyResourceChange
2024-08-20T18:04:40.236+0300 [INFO]  vcf: [SUCCESSFUL] Validate System Lock Availability on SDDC Manager: tf_resource_type=vcf_host tf_req_id=2c00dec9-6f95-50ed-6a9f-f501143a2252 tf_provider_addr=registry.terraform.io/hashicorp/vcf tf_rpc=ApplyResourceChange
2024-08-20T18:04:40.236+0300 [INFO]  vcf: [SUCCESSFUL] Update SDDC Manager Inventory with New ESXi Hosts and Associate with Given Network Pool: tf_req_id=2c00dec9-6f95-50ed-6a9f-f501143a2252 tf_provider_addr=registry.terraform.io/hashicorp/vcf tf_rpc=ApplyResourceChange tf_resource_type=vcf_host
2024-08-20T18:04:40.236+0300 [INFO]  vcf: [SUCCESSFUL] Generate SDDC Manager Acquire System Limited Lock Data: tf_req_id=2c00dec9-6f95-50ed-6a9f-f501143a2252 tf_provider_addr=registry.terraform.io/hashicorp/vcf tf_rpc=ApplyResourceChange tf_resource_type=vcf_host
2024-08-20T18:04:40.237+0300 [INFO]  vcf: [SUCCESSFUL] Acquire SDDC Manager System Limited Lock: tf_rpc=ApplyResourceChange tf_resource_type=vcf_host tf_req_id=2c00dec9-6f95-50ed-6a9f-f501143a2252 tf_provider_addr=registry.terraform.io/hashicorp/vcf
2024-08-20T18:04:40.237+0300 [INFO]  vcf: [SUCCESSFUL] Generate SDDC Manager Acquire Host(s) Lock Data: tf_req_id=2c00dec9-6f95-50ed-6a9f-f501143a2252 tf_provider_addr=registry.terraform.io/hashicorp/vcf tf_rpc=ApplyResourceChange tf_resource_type=vcf_host
2024-08-20T18:04:40.237+0300 [INFO]  vcf: [SUCCESSFUL] Acquire SDDC Manager Host(s) Lock: tf_req_id=2c00dec9-6f95-50ed-6a9f-f501143a2252 tf_provider_addr=registry.terraform.io/hashicorp/vcf tf_rpc=ApplyResourceChange tf_resource_type=vcf_host
2024-08-20T18:04:40.237+0300 [INFO]  vcf: [SUCCESSFUL] Updates SDDC Manager Task-Registration with New Resource IDs: tf_resource_type=vcf_host tf_req_id=2c00dec9-6f95-50ed-6a9f-f501143a2252 tf_provider_addr=registry.terraform.io/hashicorp/vcf tf_rpc=ApplyResourceChange
2024-08-20T18:04:40.237+0300 [INFO]  vcf: [SUCCESSFUL] Gather DNS and NTP Server Details: tf_req_id=2c00dec9-6f95-50ed-6a9f-f501143a2252 tf_provider_addr=registry.terraform.io/hashicorp/vcf tf_rpc=ApplyResourceChange tf_resource_type=vcf_host
2024-08-20T18:04:40.237+0300 [INFO]  vcf: [SUCCESSFUL] Gather New ESXi Hosts Details: tf_req_id=2c00dec9-6f95-50ed-6a9f-f501143a2252 tf_provider_addr=registry.terraform.io/hashicorp/vcf tf_rpc=ApplyResourceChange tf_resource_type=vcf_host
2024-08-20T18:04:40.237+0300 [INFO]  vcf: [SUCCESSFUL] Update SDDC Manager known_hosts with New ESXi Host SSH Keys: tf_provider_addr=registry.terraform.io/hashicorp/vcf tf_rpc=ApplyResourceChange tf_resource_type=vcf_host tf_req_id=2c00dec9-6f95-50ed-6a9f-f501143a2252
2024-08-20T18:04:40.237+0300 [INFO]  vcf: [SUCCESSFUL] Install Temporary VMCA Certificate For Host Commissioning: tf_rpc=ApplyResourceChange tf_resource_type=vcf_host tf_req_id=2c00dec9-6f95-50ed-6a9f-f501143a2252 tf_provider_addr=registry.terraform.io/hashicorp/vcf
2024-08-20T18:04:40.237+0300 [INFO]  vcf: [SUCCESSFUL] Update Input for disable ESXi lockdown mode: tf_req_id=2c00dec9-6f95-50ed-6a9f-f501143a2252 tf_provider_addr=registry.terraform.io/hashicorp/vcf tf_rpc=ApplyResourceChange tf_resource_type=vcf_host
2024-08-20T18:04:40.237+0300 [INFO]  vcf: [SUCCESSFUL] Retrieve ESXi Host Lockdown Mode Configuration: tf_req_id=2c00dec9-6f95-50ed-6a9f-f501143a2252 tf_provider_addr=registry.terraform.io/hashicorp/vcf tf_rpc=ApplyResourceChange tf_resource_type=vcf_host
2024-08-20T18:04:40.237+0300 [INFO]  vcf: [SUCCESSFUL] Disable Lockdown Mode on ESXi Hosts: tf_req_id=2c00dec9-6f95-50ed-6a9f-f501143a2252 tf_provider_addr=registry.terraform.io/hashicorp/vcf tf_rpc=ApplyResourceChange tf_resource_type=vcf_host
2024-08-20T18:04:40.237+0300 [INFO]  vcf: [SUCCESSFUL] Update Input To Create Esxi Service Accounts.: tf_req_id=2c00dec9-6f95-50ed-6a9f-f501143a2252 tf_provider_addr=registry.terraform.io/hashicorp/vcf tf_rpc=ApplyResourceChange tf_resource_type=vcf_host
2024-08-20T18:04:40.237+0300 [INFO]  vcf: [SUCCESSFUL] Create New Local ESXi User: tf_provider_addr=registry.terraform.io/hashicorp/vcf tf_rpc=ApplyResourceChange tf_resource_type=vcf_host tf_req_id=2c00dec9-6f95-50ed-6a9f-f501143a2252
2024-08-20T18:04:40.237+0300 [INFO]  vcf: [SUCCESSFUL] Assign Role to New ESXi User: tf_provider_addr=registry.terraform.io/hashicorp/vcf tf_rpc=ApplyResourceChange tf_resource_type=vcf_host tf_req_id=2c00dec9-6f95-50ed-6a9f-f501143a2252
```

**Type of Pull Request**

- [ ] This is a bug fix.
- [X] This is an enhancement or feature.
- [ ] This is a code style/formatting update.
- [ ] This is a documentation update.
- [ ] This is a refactoring update.
- [ ] This is a chore update
- [ ] This is something else.
      Please describe:

**Related to Existing Issues**

Ref: #204

**Test and Documentation Coverage**

For bug fixes or features:

- [ ] Tests have been completed.
- [ ] Documentation has been added/updated.

**Breaking Changes?**

- [ ] Yes, there are breaking changes.
- [X] No, there are no breaking changes.

### CODE REVIEW

File: internal/api_client/task_tracker.go [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+	"context"
+	"errors"
+	"fmt"
+	"github.com/hashicorp/terraform-plugin-log/tflog"
+	"github.com/vmware/terraform-provider-vcf/internal/constants"
+	"github.com/vmware/vcf-sdk-go/client"
+	"github.com/vmware/vcf-sdk-go/client/tasks"
+	"github.com/vmware/vcf-sdk-go/models"
+	"time"
+)
```
Comment (tenthirtyam) [inline internal/api_client/task_tracker.go:16]: This is more idiomatic:

```suggestion
import (
	"context"
	"errors"
	"fmt"
	"time"
	
	"github.com/hashicorp/terraform-plugin-log/tflog"
	"github.com/vmware/vcf-sdk-go/client"
	"github.com/vmware/vcf-sdk-go/client/tasks"
	"github.com/vmware/vcf-sdk-go/models"
	
	"github.com/vmware/terraform-provider-vcf/internal/constants"
)
```

File: internal/api_client/task_tracker.go [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+
+type TaskTracker struct {
+	ctx             context.Context
+	client          *client.VcfClient
+	taskId          string
+	pollingInterval time.Duration
+
+	completedTasks map[string]bool
+}
```
Comment (tenthirtyam) [inline internal/api_client/task_tracker.go:29]:
```suggestion
type TaskTracker struct {
type TaskTracker struct {
	ctx             context.Context
	client          *client.VcfClient
	taskId          string
	pollingInterval time.Duration
	completedTasks  map[string]bool
}
}
```

File: internal/api_client/task_tracker.go [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+	"github.com/vmware/terraform-provider-vcf/internal/constants"
+	"github.com/vmware/vcf-sdk-go/client"
+	"github.com/vmware/vcf-sdk-go/client/tasks"
+	"github.com/vmware/vcf-sdk-go/models"
+	"time"
+)
+
+const (
+	defaultPollingInterval = time.Second * 20
+)
```
Comment (tenthirtyam) [inline internal/api_client/task_tracker.go:20]: How about using using constants here for the status:

```go
const (
	// Default polling interval for task tracking.
	defaultPollingInterval = 20 * time.Second

	// Task status constants.
	statusInProgress    = "In Progress"
	statusPending       = "Pending"
	statusFailed        = "Failed"
	statusCancelled     = "Cancelled"
	statusCompleted     = "Completed"
	statusNotApplicable = "NOT_APPLICABLE"
)
```

File: internal/api_client/task_tracker.go [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+}
+
+func NewTaskTrackerWithCustomPollingInterval(ctx context.Context, client *client.VcfClient, taskId string, pollingInterval time.Duration) *TaskTracker {
+	tracker := NewTaskTracker(ctx, client, taskId)
+	tracker.pollingInterval = pollingInterval
+	return tracker
+}
+
+func (t *TaskTracker) WaitForTask() error {
+	ticker := time.NewTicker(t.pollingInterval)
```
Comment (tenthirtyam) [inline internal/api_client/task_tracker.go:48]:
```suggestion
	ticker := time.NewTicker(t.pollingInterval)
    defer ticker.Stop()
```

File: internal/api_client/task_tracker.go [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+				tflog.Error(t.ctx, errorMsg)
+
+				return errors.New(errorMsg)
+			}
+
+			tflog.Info(t.ctx, fmt.Sprintf("Task with ID = %s is in state %s, completed at %s",
+				task.ID, task.Status, task.CompletionTimestamp))
+			return nil
+		}
+	}
```
Comment (tenthirtyam) [inline internal/api_client/task_tracker.go:77]: You could then use the constants and a switch:

```suggestion
    for {
        select {
        case <-t.ctx.Done():
            return t.ctx.Err()
        case <-ticker.C:
            task, err := t.getTask()
            if err != nil {
                return err
            }

            t.logTask(task)

            switch task.Status {
            case statusInProgress, statusPending:
                continue
            case statusFailed, statusCancelled:
                errorMsg := fmt.Sprintf("Task with ID = %s, Name: %q, Type: %q is in state %s",
                    task.ID, task.Name, task.Type, task.Status)
                tflog.Error(t.ctx, errorMsg)
                return errors.New(errorMsg)
            default:
                tflog.Info(t.ctx, fmt.Sprintf("Task with ID = %s is in state %s, completed at %s",
                    task.ID, task.Status, task.CompletionTimestamp))
                return nil
            }
        }
    }
```

**Review by tenthirtyam — CHANGES_REQUESTED** [review]
I added some minor suggestions to refine it just a bit more.

1. **Constants**: Define constants for task statuses to avoid "magic strings" 🪄 
2. **Switch**: Use a switch statement for task status checks in WaitForTask.
3. **Context Cancellation**: Handle context cancellation in WaitForTask.

These aside, this is awesome! 🚀

File: internal/api_client/task_tracker.go (hunk 1 above)
Comment (spacegospod) [reply internal/api_client/task_tracker.go:16]: Agreed, I'll see if I can find some way to automate import formatting in my IDE.
In any case I'll fix these in the next iteration

**Review by dimitarproynov — DISMISSED** [review]
Looks good, how about the integration into the codebase?

**Comment by spacegospod** [comment]
> Looks good, how about the integration into the codebase?

I was planning to do it one resource at a time because I want to adjust the polling intervals for each individual case.
I could do it in this change but I fear that it would take forever to run the acceptance tests.
Another option is to submit it and do the integration whenever we touch a resource file.

**Comment by tenthirtyam** [comment]
> Another option is to submit it and do the integration whenever we touch a resource file.

I prefer the option to lay the groundwork here, and then create a tracking issue with each resource that we can track the updates against. For example:

- [ ] `r/cluster`
- [x] `r/domain`
- [x] `r/network_pool`
- [ ] `etc...`

**Review by tenthirtyam — APPROVED** [review]
LGTM! 🚀 

Thanks for including the recommended changes.

**Pull request merged**